### PR TITLE
Update first in RetryOnConflict

### LINKED ISSUE #1310: [Good First] Update first in RetryOnConflict
**Task description**:

Our current logic in file `pkg/controllers/execution/execution_controller.go` is to get the resource first and then update the resource. This may result in a waste of getting time. We can change it to update first, and then get if the update fails.

**Solution**:

```diff
--- a/pkg/controllers/execution/execution_controller.go
+++ b/pkg/controllers/execution/execution_controller.go
@@ -271,10 +271,20 @@ func (c *Controller) updateAppliedCondition(work *workv1alpha1.Work, status meta
        }

        return retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
-               if err = c.Get(context.TODO(), client.ObjectKey{Namespace: work.Namespace, Name: work.Name}, work); err != nil {
-                       return err
-               }
                meta.SetStatusCondition(&work.Status.Conditions, newWorkAppliedCondition)
-               return c.Status().Update(context.TODO(), work)
+               updateErr := c.Status().Update(context.TODO(), work)
+               if updateErr == nil {
+                       return nil
+               }
+
+               updated := &workv1alpha1.Work{}
+               if err = c.Get(context.TODO(), client.ObjectKey{Namespace: work.Namespace, Name: work.Name}, updated); err == nil {
+                       // make a copy, so we don't mutate the shared cache
+                       work = updated.DeepCopy()
+               } else {
+                       klog.Errorf("failed to get updated work %s/%s: %v", work.Namespace, work.Name, err)
+               }
+
+               return updateErr
        })
 }
```

**Who can join or take the task**:

The good first issue is intended for `first-time contributors` to get started on his/her contributor journey.

After a contributor has successfully completed 1-2 good first issue's,
they should be ready to move on to `help wanted` items, saving the remaining `good first issue` for other new contributors.

**How to join or take the task**:

Just reply on the issue with the message `/assign` in a separate line.

Then, the issue will be assigned to you.

**How to ask for help**:

If you need help or have questions, please feel free to ask on this issue.
The issue author or other members of the community will guide you through the contribution process.

### PR DESCRIPTION
**Pull request by RishiKumarRay**
**What type of PR is this?**
Update first in RetryOnConflict
**Does this PR introduce a user-facing change?:**
NONE
Add one of the following kinds:

/kind feature

**What this PR does / why we need it**:
Our current logic in file pkg/controllers/execution/execution_controller.go is to get the resource first and then update the resource. This may result in a waste of getting time. We can change it to update first, and then get if the update fails.

**Which issue(s) this PR fixes**:
Fixes #1310 
```release-note
None
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Hi @RishiKumarRay, maybe you need to sign off refer to https://github.com/karmada-io/karmada/pull/1311/checks?check_run_id=4933829645

**Comment by RishiKumarRay** [comment]
@XiShanYongYe-Chang yes i have updated that, thanks

**Comment by XiShanYongYe-Chang** [comment]
> **What type of PR is this?** Update first in RetryOnConflict Add one of the following kinds:
> 
> /kind feature
> 
> **What this PR does / why we need it**: Our current logic in file pkg/controllers/execution/execution_controller.go is to get the resource first and then update the resource. This may result in a waste of getting time. We can change it to update first, and then get if the update fails.
> 
> **Which issue(s) this PR fixes**: Fixes #1310

Add the following information in the PR description:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Comment by RishiKumarRay** [comment]
@XiShanYongYe-Chang updated that too, thanks

**Comment by RainbowMango** [comment]
Hi @RishiKumarRay 
You might have to solve the lint issue:
```
pkg/controllers/execution/execution_controller.go:274: File is not `gofmt`-ed with `-s` (gofmt)
		
pkg/controllers/execution/execution_controller.go:273: unnecessary leading newline (whitespace)
	return retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
```
https://github.com/karmada-io/karmada/runs/4933942294?check_suite_focus=true

**Comment by RishiKumarRay** [comment]
@RainbowMango Thanks, I have updated , i think few whitespaces were there but i have fix them , can you check now please

**Review by RainbowMango — COMMENTED** [review]
LGTM

Could you please [squash your commits](https://github.com/karmada-io/karmada/blob/master/docs/contributors/guide/github-workflow.md#squash-commits)?

**Comment by RishiKumarRay** [comment]
@RainbowMango i have squashed the commit, can you please confirm if it is right?

File: cmd/agent/app/options/options.go [hunk 1]
```diff
@@ -93,4 +96,5 @@ func (o *Options) AddFlags(fs *pflag.FlagSet, allControllers []string) {
 	fs.DurationVar(&o.ClusterCacheSyncTimeout.Duration, "cluster-cache-sync-timeout", util.CacheSyncTimeout, "Timeout period waiting for cluster cache to sync.")
 	fs.StringVar(&o.ClusterAPIEndpoint, "cluster-api-endpoint", o.ClusterAPIEndpoint, "APIEndpoint of the cluster.")
 	fs.StringVar(&o.ProxyServerAddress, "proxy-server-address", o.ProxyServerAddress, "Address of the proxy server that is used to proxy to the cluster.")
+	fs.DurationVar(&o.ResyncPeriod.Duration, "resync-period", 0, "Base frequency the informers are resynced.")
```
[the remark this answers is not in the document]
Comment (RishiKumarRay) [reply cmd/agent/app/options/options.go]: @RainbowMango Thank you , i have done this too, sorry i am new to this so i was not aware about this , From next time i will make sure i follow all this

[the remark this answers is not in the document]
Comment (RishiKumarRay) [reply cmd/agent/app/options/options.go]: @RainbowMango Yeah i used the master branch , i will make sure not to repeat this again from next time. I am still learning

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

It will be all set now and will be merged automatically after all tests done.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm

**Comment by XiShanYongYe-Chang** [comment]
Hi, @RishiKumarRay , the CI lint failed, you can use `go fmt` for setting.

**Comment by RishiKumarRay** [comment]
@XiShanYongYe-Chang how to use go fmt?

**Comment by XiShanYongYe-Chang** [comment]
> how to use go fmt

We can refer to https://go.dev/blog/gofmt, if you use goland, please refer to https://stackoverflow.com/questions/47735678/goland-how-to-use-gofmt

**Comment by RishiKumarRay** [comment]
@XiShanYongYe-Chang Can you please check this now, i have updated

**Comment by XiShanYongYe-Chang** [comment]
@RishiKumarRay It works ok~!

/lgtm

**Comment by RishiKumarRay** [comment]
@RainbowMango @XiShanYongYe-Chang Thank you for your guidance , It is merged now.

**Comment by RainbowMango** [comment]
Congrats!!! and welcome!!! @RishiKumarRay

**Pull request merged**